src/WebJobs.Script.WebHost/Metrics/LinuxContainerMetricsPublisher.cs

```diff
@@ -0,0 +1,235 @@
+            _environment = environment ?? throw new ArgumentNullException(nameof(environment));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _hostNameProvider = hostNameProvider ?? throw new ArgumentNullException(nameof(hostNameProvider));
+            _memoryActivities = new BlockingCollection<MemoryActivity>(new ConcurrentQueue<MemoryActivity>(), boundedCapacity: _maxBufferSize);
+            _functionActivities = new BlockingCollection<FunctionActivity>(new ConcurrentQueue<FunctionActivity>(), boundedCapacity: _maxBufferSize);
+
+            _currentMemoryActivities = new ConcurrentQueue<MemoryActivity>();
+            _currentFunctionActivities = new ConcurrentQueue<FunctionActivity>();
+
+            var nodeIpAddress = _environment.GetEnvironmentVariable(EnvironmentSettingNames.LinuxNodeIpAddress);
+            _requestUri = string.Format(_requestUriFormat, nodeIpAddress);
+
+            _containerName = _environment.GetEnvironmentVariable(EnvironmentSettingNames.ContainerName);
+
+            if (_standbyOptions.CurrentValue.InStandbyMode)
+            {
+                _standbyOptionsOnChangeSubscription = _standbyOptions.OnChange(o => OnStandbyOptionsChange());
+            }
+            else
+            {
+                Start();
+            }
+        }
+
+        private void OnStandbyOptionsChange()
+        {
+            if (!_standbyOptions.CurrentValue.InStandbyMode)
+            {
+                Start();
+            }
+        }
+
+        public void AddFunctionExecutionActivity(string functionName, string invocationId, int concurrency, string executionStage, bool success, long executionTimeSpan, DateTime eventTimeStamp)
+        {
+            FunctionActivity activity = new FunctionActivity
+            {
+                FunctionName = functionName,
+                InvocationId = invocationId,
+                Concurrency = concurrency,
+                ExecutionStage = executionStage,
+                IsSucceeded = success,
+                ExecutionTimeSpanInMs = executionTimeSpan,
+                EventTimeStamp = eventTimeStamp,
+                Tenant = _tenant
+            };
+
+            if (!_functionActivities.TryAdd(activity))
+            {
+                _logger.LogWarning("Buffer full.Dropping current batch of function activities");
+                DrainActivities(_currentFunctionActivities, _functionActivities);
+            }
+
+            _logger.LogInformation($"Added function activity : {functionName} {invocationId} {concurrency} {executionStage} {success} {executionTimeSpan}");
+        }
+
+        public void AddMemoryActivity(DateTime timeStampUtc, long data)
+        {
+            var memoryActivity = new MemoryActivity
+            {
+                CommitSizeInBytes = data,
+                EventTimeStamp = timeStampUtc,
+                Tenant = _tenant
+            };
+
+            if (!_memoryActivities.TryAdd(memoryActivity))
```

Comment:
I think this code would be simpler if you just used a single collection for each activity type, not the dual collection with copy operations that you're doing now. Just like we do in FileWriter.cs - we just have a single buffer we're writing to, and when we flush we just snapshot the current contents for publish within a lock. Same problem as here.

Also in FileWriter, when we hit limit we don't have any drain/copy operations to perform - we just stop taking new entries. Why when you hit the limit here do you drop all the _previous_ logs rather than the new one coming in?

Reply:
Because it is possible something is wrong with the current batch of events we are trying to send. In those cases, we just want to skip the current batch and pick up the next one. The other scenario is that metrics publisher fails because the metering service is down for an extended period of time. In those cases as well, we should drop the older events rather than the new ones.